extensions/binding/org.eclipse.smarthome.binding.mqtt.generic/src/main/java/org/eclipse/smarthome/binding/mqtt/generic/internal/NumberValue.java

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.binding.mqtt.generic.internal;
+
+import java.math.BigDecimal;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.CoreItemFactory;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.IncreaseDecreaseType;
+import org.eclipse.smarthome.core.library.types.PercentType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.library.types.UpDownType;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.State;
+
+/**
+ * Implements a percentage value. Minimum and maximum are definable.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@NonNullByDefault
+public class NumberValue implements AbstractMqttThingValue {
+    private final double min;
+    private final double max;
+    private final double step;
+    private final Boolean isFloat;
+    private final boolean isPercent;
+
+    private PercentType percentValue;
+    private DecimalType numberValue;
+
+    public NumberValue(@Nullable Boolean isFloat, @Nullable BigDecimal min, @Nullable BigDecimal max,
+            @Nullable BigDecimal step, boolean isPercent) {
+        this.isFloat = isFloat == null ? true : isFloat;
+        this.min = min == null ? 0.0 : min.doubleValue();
+        this.max = max == null ? 100.0 : max.doubleValue();
+        this.step = step == null ? 1.0 : step.doubleValue();
+        this.isPercent = isPercent;
+        percentValue = new PercentType();
+        numberValue = new DecimalType();
+    }
+
+    @Override
+    public State getValue() {
+        return isPercent ? percentValue : numberValue;
+    }
+
+    @Override
+    public String update(Command command) throws IllegalArgumentException {
+        if (isPercent) {
+            if (command instanceof StringType) {
+                double v = Double.valueOf(((StringType) command).toString());
+                v = (v - min) * 100.0 / (max - min);
```

Comment:
Min,Max might be checked in the constructor and a `ConfigurationException` or similar can be thrown, true.

Reply:
Do you want to guard v als against having a value below min or above max? You can use Math.min and Math.max. These methods can also be used below, instead of your own checks.

Reply:
Do you want to guard v als against having a value below min or above max? You can use Math.min and Math.max. These methods can also be used below, instead of your own checks.